src/Symfony/Component/Security/Http/Authenticator/FormLoginAuthenticator.php

```diff
@@ -123,16 +123,16 @@ private function getCredentials(Request $request): array
             $credentials['password'] = ParameterBagUtils::getRequestParameterValue($request, $this->options['password_parameter']) ?? '';
         }
 
-        if (!\is_string($credentials['username']) && !$credentials['username'] instanceof \Stringable) {
-            throw new BadRequestHttpException(sprintf('The key "%s" must be a string, "%s" given.', $this->options['username_parameter'], \gettype($credentials['username'])));
+        if (!(\is_string($credentials['username']) || $credentials['username'] instanceof \Stringable) || '' === $credentials['username']) {
```

Comment:
this implementation is flawed:

- it does not check for Stringable being empty strings after casting
- the logic just below trims the username. The check for empty string should probably be done after trimming rather than before otherwise, submitting some spaces can still produce an empty username)

Reply:
I moved the check for empty after `trim` and in case of stringable object I also added a mandatory string cast. I hope this more to what you had in mind? 

The message I copied from JsonLoginAuthenticator.